apps/liteprotocoltester/statistics.nim

```diff
@@ -0,0 +1,173 @@
+
+proc addMessage*(self: var Statistics, msg: ProtocolTesterMessage) =
+  if self.allMessageCount == 0:
+    self.allMessageCount = msg.count
+  elif self.allMessageCount != msg.count:
+    warn "Message count mismatch at message",
+      index = msg.index, expected = self.allMessageCount, got = msg.count
+
+  if not self.helper.seenIndicies.contains(msg.index):
+    self.helper.seenIndicies.incl(msg.index)
+  else:
+    inc(self.duplicateCount)
+    warn "Duplicate message", index = msg.index
+    ## just do not count into stats
+    return
+
+  ## detect misorder arrival and possible lost messages
+  if self.helper.prevIndex + 1 < msg.index:
+    inc(self.misorderCount)
+    warn "Misordered message arrival",
+      index = msg.index, expected = self.helper.prevIndex + 1
+
+    ## collect possible lost message indicies
+    for idx in self.helper.prevIndex + 1 ..< msg.index:
+      self.helper.lostIndicies.incl(idx)
+
+  ## may remove late arrival
+  self.helper.lostIndicies.excl(msg.index)
+
+  ## calculate latency
+  let currentArrivedAt = Moment.now()
+  let delaySincePrevArrived = self.helper.prevArrivedAt - currentArrivedAt
+
+  let expectedDelay = nanos(msg.sincePrev)
+
+  let latency = delaySincePrevArrived - expectedDelay
+  self.helper.prevArrivedAt = currentArrivedAt
+  self.helper.prevIndex = msg.index
+
+  inc(self.receivedMessages)
+
+  if self.minLatency.isZero or (latency < self.minLatency and latency > nanos(0)):
+    self.minLatency = latency
+  if latency > self.maxLatency:
+    self.maxLatency = latency
+
+  self.cummulativeLatency += latency
+
+proc addMessage*(
+    self: var PerPeerStatistics, peerId: string, msg: ProtocolTesterMessage
+) =
+  if not self.contains(peerId):
+    self[peerId] = Statistics.init()
+
+  discard catch:
+    self[peerId].addMessage(msg)
+
+proc lossCount*(self: Statistics): uint32 =
+  self.allMessageCount - self.receivedMessages
+
+proc averageLatency*(self: Statistics): Duration =
+  if self.receivedMessages == 0:
+    return nanos(0)
+  return self.cummulativeLatency div self.receivedMessages
+
+proc echoStat*(self: Statistics) =
+  let printable = catch:
+    """*----------------------------------------------------------------*
+|  Expected  |  Reveived  |    Loss    |  Misorder  |  Duplicate |
+|{self.allMessageCount:>11} |{self.receivedMessages:>11} |{self.lossCount():>11} |{self.misorderCount:>11} |{self.duplicateCount:>11} |
+*----------------------------------------------------------------*
+| Latency stat:                                                  |
+|    avg latency: {self.averageLatency():<47}|
+|    min latency: {self.maxLatency:<47}|
+|    max latency: {self.minLatency:<47}|
+*----------------------------------------------------------------*""".fmt()
+
+  if printable.isErr:
+    echo "Error while printing statistics"
+  else:
+    echo printable.get()
+
+proc jsonStat*(self: Statistics): string =
+  let json = catch:
+    """{{\"expected\":{self.allMessageCount},
```

Comment:
Are the `\` needed when using the triple-quote?

Reply:
Yes, they are not needed!